api/app.py

```
@@ -10,67 +13,257 @@
 def create_recipe():
-    new_recipe = {
-        'id': len(recipes) + 1,
-        'name': request.json.get('name'),
-        'duration': request.json.get('duration'),
-        'pictures': request.json.get('pictures'),
-        'instructions': request.json.get('instructions'),
-        'categories': request.json.get('categories'),
-        'ingredients': request.json.get('ingredients'),
-    }
-    recipes.append(new_recipe)
-    return jsonify(new_recipe)
+    data = request.get_json()
+    if data is None:
+        return jsonify({'message': 'No input data provided'}), 400
+
+    name = data.get('name')
+    if not name  or not isinstance(name, str) or name.strip() == "":
+        return jsonify({'message': 'Name is required'}), 400
+
+    duration = data.get('duration')
+    if not duration or not isinstance(duration, str) or duration.strip() == "":
+        return jsonify({'error': 'Duration must be a non-empty string'}), 400
+
+    pictures = data.get('pictures')
+    if not pictures or not isinstance(pictures, list) or not all(isinstance(pic, str) for pic in pictures):
+        return jsonify({'message': 'Pictures are required as a list of valid URLs'}), 400
+
+    instructions = data.get('instructions')
+    if not instructions or not isinstance(instructions, str) or instructions.strip() == "":
+        return jsonify({'message': 'Instructions are required'}), 400
+
+    categories = []
+    categories_data = data.get('categories', [])
+    for category in categories_data:
+        if isinstance(category, str):
+            name = category
+            color = '#808080'
+        elif isinstance(category, dict):
+            name = category['name']
+            color = category.get('color', '#808080')
+        else:
+            return jsonify({'message': 'Invalid format'}), 400
+        if not name:
+            jsonify({'message': 'No name provided'}), 400
+        cat = Category.query.filter_by(name=name).first()
+        if not cat:
```

We don't need to create a `category` that doesn't exist here; we'll use the `create_category()` endpoint. Here you can return an error message if the category does not exist, for example:
```python
return jsonify({'error': f'category {cat["name"]} not found'}), 400
``` 
Same for `edit_recipe()`.